specs/simple-serialize.md

```diff
@@ -1,424 +1,114 @@
-```
-
-### Tree Hash
-
-The below `hash_tree_root_internal` algorithm is defined recursively in the case of lists and containers, and it outputs a value equal to or less than 32 bytes in size. For use as a "final output" (eg. for signing), use `hash_tree_root(x) = zpad(hash_tree_root_internal(x), 32)`, where `zpad` is a helper that extends the given `bytes` value to the desired `length` by adding zero bytes on the right:
-
-```python
-def zpad(input: bytes, length: int) -> bytes:
-   return input + b'\x00' * (length - len(input))
+assert value in (True, False)
+return b'\x01' if value is True else b'\x00'
 ```
 
-Refer to [the helper function `hash`](https://github.com/ethereum/eth2.0-specs/blob/master/specs/core/0_beacon-chain.md#hash) of Phase 0 of the [Eth2.0 specs](https://github.com/ethereum/eth2.0-specs) for a definition of the hash function used below, `hash(x)`.
-
-#### `uint8`..`uint256`, `bool`, `bytes1`..`bytes32`
-
-Return the serialization of the value.
-
-#### `uint264`..`uintN`, `bytes33`..`bytesN`
-
-Return the hash of the serialization of the value.
-
-#### List/Vectors
-
-First, we define the Merkle tree function.
+### Containers, tuples, lists
 
 ```python
-# Merkle tree hash of a list of homogenous, non-empty items
-def merkle_hash(lst):
-    # Store length of list (to compensate for non-bijectiveness of padding)
-    datalen = len(lst).to_bytes(32, 'little')
-
-    if len(lst) == 0:
-        # Handle empty list case
-        chunkz = [b'\x00' * SSZ_CHUNK_SIZE]
-    elif len(lst[0]) < SSZ_CHUNK_SIZE:
-        # See how many items fit in a chunk
-        items_per_chunk = SSZ_CHUNK_SIZE // len(lst[0])
-
-        # Build a list of chunks based on the number of items in the chunk
-        chunkz = [
-            zpad(b''.join(lst[i:i + items_per_chunk]), SSZ_CHUNK_SIZE) 
-            for i in range(0, len(lst), items_per_chunk)
-        ]
-    else:
-        # Leave large items alone
-        chunkz = lst
-
-    # Merkleise
-    def next_power_of_2(x):  
-        return 1 if x == 0 else 2**(x - 1).bit_length()
-
-    for i in range(len(chunkz), next_power_of_2(len(chunkz))):
-        chunkz.append(b'\x00' * SSZ_CHUNK_SIZE)
-    while len(chunkz) > 1:     
-        chunkz = [hash(chunkz[i] + chunkz[i+1]) for i in range(0, len(chunkz), 2)]
-
-    # Return hash of root and data length
-    return hash(chunkz[0] + datalen)
+serialized_bytes = ''.join([serialize(element) for element in value])
+assert len(serialized_bytes) < 2**(8 * BYTES_PER_LENGTH_PREFIX)
+serialized_length = len(serialized_bytes).to_bytes(BYTES_PER_LENGTH_PREFIX, 'little')
+return serialized_length + serialized_bytes
 ```
 
-To `hash_tree_root_internal` a list, we simply do:
+## Deserialization
 
-```python
-return merkle_hash([hash_tree_root_internal(item) for item in value])
-```
+Given a type, serialization is an injective function from objects of that type to byte strings. That is, deserialization—the inverse function—is well-defined.
```

Comment:
This sounds confusing and not sure how many people will get what this means.

Maybe something like "Because serialization is an injective function (ie. if two objects of the same type are not exactly identical they will serialize to different values), for any bytestring there will always be at most one object of any given type that it could deserialize to (or zero if a bytestring is not valid SSZ). An efficient algorithm for computing this object can be found in the implementation [here](http://github.com/fill/in/the/correct/link)."